pkg/workloads/console/runner/runner.go

```diff
@@ -0,0 +1,162 @@
+type Options struct {
+	Cmd     []string
+	Timeout int
+
+	// TODO: For now we assume that all consoles are interactive, i.e. we setup a TTY on
+	// them when spawning them. This does not enforce a requirement to attach to the console
+	// though.
+	// Later on we may need to implement non-interactive consoles, for processes which
+	// expect a TTY to not be present?
+	// However with these types of consoles it will not be possible to send input to them
+	// when reattaching, e.g. attempting to send a SIGINT to cancel the running process.
+	// Interactive bool
+}
+
+// New builds a runner, with a given clientset
+func New(clientset versioned.Interface) *Runner {
+	return &Runner{
+		clientset: clientset,
+	}
+}
+
+// Create builds a console according to the supplied options and submits it to the API
+func (c *Runner) Create(namespace string, template workloadsv1alpha1.ConsoleTemplate, opts Options) (*workloadsv1alpha1.Console, error) {
+	csl := &workloadsv1alpha1.Console{
+		ObjectMeta: metav1.ObjectMeta{
+			// Let Kubernetes generate a unique name
+			GenerateName: template.Name + "-",
+
+			// TODO: Consider which labels and annotations we might want to set on the console
+			// itself.
+			// Labels:
+			// Annotations:
+		},
+		Spec: workloadsv1alpha1.ConsoleSpec{
+			ConsoleTemplateRef: corev1.LocalObjectReference{Name: template.Name},
+			// If the flag is not provided then the value will default to 0. The controller
+			// should detect this and apply the default timeout that is defined in the template.
+			TimeoutSeconds: opts.Timeout,
+			// TODO: This field is currently missing from the console spec
+			// Command:            c.Cmd,
+		},
+	}
+
+	client := c.clientset.WorkloadsV1alpha1().Consoles(namespace)
+	return client.Create(csl)
+}
+
+// FindTemplateBySelector will search for a template matching the given label
+// selector and return errors if none or multiple are found (when the selector
+// is too broad)
+func (c *Runner) FindTemplateBySelector(namespace string, labelSelector string) (*workloadsv1alpha1.ConsoleTemplate, error) {
+	client := c.clientset.WorkloadsV1alpha1().ConsoleTemplates(namespace)
+
+	templates, err := client.List(
+		metav1.ListOptions{
+			LabelSelector: labelSelector,
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list consoles templates")
+	}
+
+	if len(templates.Items) != 1 {
+		identifiers := []string{}
+		for _, item := range templates.Items {
+			identifiers = append(identifiers, item.Namespace+"/"+item.Name)
+		}
+
+		return nil, errors.Errorf(
+			"expected to discover 1 console template, but actually found: %s",
+			identifiers,
+		)
+	}
+
+	template := templates.Items[0]
+
+	return &template, nil
+}
+
+// WaitUntilReady will block until the console reaches a phase that indicates
+// that it's ready to be attached to, or has failed.
+func (c *Runner) WaitUntilReady(ctx context.Context, createdCsl workloadsv1alpha1.Console) error {
```

Comment:
This whole function is a bit ugly. It should probably return machine-parseable errors, rather than error strings, and de-duplicate some of the state checking logic.

However I got a bit sick of writing tests for it at this point, and it does do the job we need for the moment, so would prefer to try and get it merged and continue work from there if possible!

Reply:
I reckon this becomes a lot simpler if it returns `(*ConsolePhase, error)`, with the expectation it'll try to return `ConsoleRunning` if it can. Then you'd use it like this
```go
phase, err := runner.WaitUntilReady(..)
if err != nil {
 ...
}
if phase != workloadsv1alpha1.ConsoleRunning {
 ...
}
// now do your stuff
```

Then you only return errors if you genuinely can't fetch the console phase for some reason - and otherwise you just return the phase. Happy to leave this for a followup though.

Reply:
Yeah I think something like this is probably the way to go! I've made a note of it though and think that we should rework afterwards